### PR TITLE
Fix debug messages and flaws which were shown if a meeting does not exist

### PR DESCRIPTION
**Pull request by abias**
This patch fixes debug messages and flaws which were introduced in commit c22c108 / issue #207 and which were shown if a meeting did not exist on Zoom anymore but the activity overview page showed the unavailability note anyway.

With this patch, the unavailability note is only shown if the meeting exists on Zoom, i.e. the recreate message is not shown.
It also changes the implementation of the recreate message on-the-fly to use the constants from the notification class.

### CODE REVIEW

File: view.php [hunk 1]
```diff
@@ -128,10 +130,10 @@
     // Only show recreate/delete links in the message for users that can edit.
     if ($iszoommanager) {
         $message = get_string('zoomerr_meetingnotfound', 'mod_zoom', zoom_meetingnotfound_param($cm->id));
-        $style = 'notifywarning';
+        $style = \core\output\notification::NOTIFY_ERROR;
     } else {
         $message = get_string('zoomerr_meetingnotfound_info', 'mod_zoom');
-        $style = 'notifymessage';
+        $style = \core\output\notification::NOTIFY_WARNING;
```
Comment (jrchamp) [inline view.php:136]: @abias Is the switch from `'notifymessage'`/`NOTIFY_INFO` to `NOTIFY_WARNING` intentional? By default, `'notifymessage'` is mapped to `NOTIFY_INFO`. I could probably be convinced that either option is appropriate.

https://github.com/moodle/moodle/blob/18aafd0ed4d89447cf7a6ac54c8aa7b710299563/lib/outputrenderers.php#L2909-L2916
```php
            // Legacy types mapped to current types.
            'notifyproblem'     => \core\output\notification::NOTIFY_ERROR,
            'notifytiny'        => \core\output\notification::NOTIFY_ERROR,
            'notifyerror'       => \core\output\notification::NOTIFY_ERROR,
            'notifysuccess'     => \core\output\notification::NOTIFY_SUCCESS,
            'notifymessage'     => \core\output\notification::NOTIFY_INFO,
            'notifyredirect'    => \core\output\notification::NOTIFY_INFO,
            'redirectmessage'   => \core\output\notification::NOTIFY_INFO,
```

@andrewnicols Is there a reason that notifywarning is not in the legacy type map? It looks like there's still one use of it in core as part of search: https://github.com/moodle/moodle/blob/87afa4d7595eeafdf9e3f83d53272ee61fb66780/search/index.php#L181

File: view.php [hunk 2]
```diff
@@ -83,15 +83,17 @@
...
-$hostmoodleuser = new stdClass();
-$hostmoodleuser->firstname = $hostuser->first_name;
-$hostmoodleuser->lastname = $hostuser->last_name;
-$hostmoodleuser->alternatename = '';
-$hostmoodleuser->firstnamephonetic = '';
-$hostmoodleuser->lastnamephonetic = '';
-$hostmoodleuser->middlename = '';
-
-$meetinginvite = $service->get_meeting_invitation($zoom);
+if (!$showrecreate) {
```
Comment (jrchamp) [inline view.php:86]: It seems like this the first chunk of code depends on $hostuser instead of $showrecreate:
```suggestion
if ($hostuser) {
```

**Comment by jrchamp** [comment]
$hostmoodleuser is also used here: https://github.com/ucla/moodle-mod_zoom/blob/a8af575a1f0fbf132a235fdac5c9716cff485c86/view.php#L165

Should there be a check in this section of code also?

File: view.php (hunk 1 above)
Comment (abias) [reply view.php:136]: @jrchamp , yes the switch from NOTIFY_INFO to NOTIFY_WARNING was intentional by me. I just didn't mention it in the commit message, sorry. 

My thoughts were that NOTIFY_INFO is just displayed with a grey background, making it quite difficult for the student to see at first glance that the meeting instance is not functional at all and that there will never appear a "Join meeting" button.


Furthermore, I am unsure if the ping to Andrew worked if he has never contributed anything to this repo. Maybe you want to ask him in Moodle Dev chat again.

File: view.php (hunk 2 above)
Comment (abias) [reply view.php]: I think this change make any functional difference, but it is semantically more understandable.

I have pushed an improvement on https://github.com/ucla/moodle-mod_zoom/pull/245/commits/4d8f3a6dbcff4053978818325e85cd440a894546

**Comment by abias** [comment]
> $hostmoodleuser is also used here:
> 
> https://github.com/ucla/moodle-mod_zoom/blob/a8af575a1f0fbf132a235fdac5c9716cff485c86/view.php#L165
> 
> Should there be a check in this section of code also?

@jrchamp , if you look further up on line https://github.com/ucla/moodle-mod_zoom/blob/a8af575a1f0fbf132a235fdac5c9716cff485c86/view.php#L150, there is already such a check. Isn't this sufficient?

**Comment by jrchamp** [comment]
@abias $hostuser is an object of host user information https://github.com/ucla/moodle-mod_zoom/blob/a8af575a1f0fbf132a235fdac5c9716cff485c86/view.php#L80

So `if ($hostuser)` answers the question "Does this script have an object that contains the host user's information?"
`if ($userishost == true)` answers the question "Is the current user a host for this meeting?"

These are very different checks, so it's possible for the $hostmoodleuser object does not exist when we go to use it for the capacity warning. Overall, it seems like the capacity warning might not be all that helpful when $showrecreate is true, because we will have already printed out a notification that the meeting doesn't exist.

Does this clear up the discussion?

File: view.php (hunk 1 above)
Comment (jrchamp) [reply view.php:139]: That all makes sense. I really like your thought process for upgrading those error messages.

**Comment by abias** [comment]
Hi @jrchamp ,

thank you for the additional notes.

I have now rebased the branch onto the latest master branch, have added another check which prevents the capacity warning from being processed when $showrecreate is true and have squashed the commits.

I would be grateful if you could have another look and tell me if there is anything which I have missed in the discussion or which you would want to have changed.

Thanks,
Alex

**Comment by jrchamp** [comment]
Thank you! This is very easy to read now, which is why I'm now noticing a potential enhancement.

Near the top, we perform an API call to get the `$meetinginvite`:
https://github.com/ncstate-delta/moodle-mod_zoom/blob/ab6dd4fe10fe0aafacc2d30b28e4a33a4c98c16b/view.php#L96-L99

Near the bottom, we may or may not use the `$meetinginvite`:
https://github.com/ncstate-delta/moodle-mod_zoom/blob/ab6dd4fe10fe0aafacc2d30b28e4a33a4c98c16b/view.php#L403-L413

In an effort to reduce the unnecessary API calls, it seems beneficial to merge these two sections and prioritize the least expensive tests. Proposal:
```php
// Show dial-in information. 
if (!$showrecreate
    && ($zoom->option_audio === ZOOM_AUDIO_BOTH || $zoom->option_audio === ZOOM_AUDIO_TELEPHONY)
    && ($userishost || has_capability('mod/zoom:viewdialin', $context))) {
    // Get meeting invitation from Zoom.
    $meetinginvite = $service->get_meeting_invitation($zoom)->get_display_string($cm->id);
    if (!empty($meetinginvite)) {
        $meetinginvitetext = str_replace("\r\n", '<br/>', $meetinginvite); 
        $showbutton = html_writer::tag('button', $strmeetinginviteshow, 
            array('id' => 'show-more-button', 'class' => 'btn btn-link pt-0 pl-0')); 
        $meetinginvitebody = html_writer::div($meetinginvitetext, '', 
            array('id' => 'show-more-body', 'style' => 'display: none;')); 
        $table->data[] = array($strmeetinginvite, html_writer::div($showbutton . $meetinginvitebody, '')); 
    }
}
```

What are your thoughts on this proposal?

**Comment by abias** [comment]
That absolutely makes sense, @jrchamp .

I have updated and force-pushed the branch with your proposal.

Cheers,
Alex

**Review by jrchamp — DISMISSED** [review]
This looks great! Thanks @abias

**Comment by abias** [comment]
Thanks for the feedback, @jrchamp .

Now that the PR is approved by you and as the repo was also transferred to your organization some days ago, are you also going to be the one to merge it or will this still be @rlorenzo ?

**Comment by jrchamp** [comment]
Good question! I'm not sure how our Moodle team plans to handle these, but I expect that you'll see more involvement from @smbader once our May maintenance is complete. There are several areas where I would be delighted to have @rlorenzo's continued involvement, but time changes all things and the community has already been given so much by their contributions.

**Comment by abias** [comment]
Hi @jrchamp ,

thanks for your answer. 
Maybe you misunderstood my question, I didn't intend to ask for the 'big picture' here. But it's interesting to get some insights from you.

Let me rephrase the question: As far as I see, the issues and hickups with this PR are solved and the PR is now ready to be merged. So, can someone please merge it so that we don't pile up rebase effort in the other open PRs?

Thanks,
Alex

**Comment by abias** [comment]
I have rebased this branch onto the latest master branch and have force-pushed the branch.

**Review by jrchamp — APPROVED** [review]
Looks great!

Brief summary of the code:

- Only fetch meeting invitation when it will be used.
- Only create `$hostmoodleuser` object when `$hostuser` object is populated.
- Only show warning messages when the higher-priority "recreate" message is not displayed.
- Use current notification constants that were added by Moodle in Moodle 2.9. (This is safe because we already require Moodle 3.3 or newer via version.php 2017051500)

**Review by smbader — APPROVED** [review]
Reviewed with @jrchamp, changes look good.
